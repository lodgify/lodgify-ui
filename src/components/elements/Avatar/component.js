import React from 'react';
import { string } from 'prop-types';

import { Thumbnail } from 'media/Thumbnail';
import { testidFactory } from 'utils/testid';

const testid = testidFactory('avatar');

export const Component = ({ firstname, lastname, image }) => (
  <div className="avatar" data-testid={testid()}>
    {!!image ? (
      <Thumbnail data-testid={testid('image')} imageUrl={image} isCircular />
    ) : (
      <p data-testid={testid('text')}>
        {`${firstname[0]}${lastname[0]}`.toUpperCase()}
      </p>
    )}
  </div>
);

Component.displayName = 'Avatar';
Component.defaultProps = {
  firstname: ' ',
  lastname: ' ',
  image: null,
};
Component.propTypes = {
  firstname: string,
  image: string,
  lastname: string,
};
