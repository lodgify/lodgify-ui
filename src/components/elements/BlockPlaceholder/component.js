import React from 'react';
import { PropTypes } from 'prop-types';
import classnames from 'classnames';

/**
 * A BlockPlaceholder indicates that a block of content will soon appear in a layout.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ isFluid, isRectangular, isSquare }) => (
  <div
    className={classnames('placeholder block', {
      fluid: isFluid,
      rectangle: isRectangular,
      square: isSquare,
    })}
  />
);

Component.displayName = 'BlockPlaceholder';

Component.defaultProps = {
  isFluid: false,
  isSquare: false,
  isRectangular: false,
};

Component.propTypes = {
  /** The placeholder fills the height and width of its parent. */
  isFluid: PropTypes.bool,
  /** The placeholder maintains 4:3 proportions. */
  isRectangular: PropTypes.bool,
  /** The placeholder maintains 1:1 proportions. */
  isSquare: PropTypes.bool,
};
