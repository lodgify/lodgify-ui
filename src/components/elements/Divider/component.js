import React from 'react';
import PropTypes from 'prop-types';
import { Divider } from 'semantic-ui-react';
import getClassNames from 'classnames';

import { getIsSizeLarge } from './utils/getIsSizeLarge';

/**
 * A divider adds whitespace between other elements.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ className, hasLine, size }) => (
  <Divider
    className={getClassNames(className, {
      'is-size-small': size === 'small',
      'is-size-huge': size === 'huge',
    })}
    hidden={!hasLine}
    section={getIsSizeLarge(size)}
  />
);

Component.displayName = 'Divider';

Component.defaultProps = {
  className: '',
  hasLine: false,
  size: 'medium',
};

Component.propTypes = {
  /**
   * Custom class name.
   * Provided by `ShowOn` so ignored in the styleguide.
   * @ignore
   */
  className: PropTypes.string,
  /** Does the divider have a visible line. */
  hasLine: PropTypes.bool,
  /** The size of the divider. */
  size: PropTypes.oneOf(['small', 'medium', 'large', 'huge']),
};
