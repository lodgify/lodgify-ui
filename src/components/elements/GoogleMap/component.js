import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Card } from 'semantic-ui-react';
import { debounce } from 'debounce';

import { ReactGoogleMap } from 'utils/react-google-maps';

import { getParentNodeWidth } from './utils/getParentNodeWidth';
import { getHeight } from './utils/getHeight';
import { getMaxWidth } from './utils/getMaxWidth';
import { getImgSrc } from './utils/getImgSrc';
import { GOOGLE_MAPS_API_KEY } from './constants';

/**
 * A map displays a location.
 * By default, the map will display a static image then load a dynamic interface on user interaction.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export class Component extends PureComponent {
  state = {
    isDynamicMapShowing: false,
    isMouseOverMap: false,
    parentNodeWidth: 0,
  };

  componentDidMount = () => {
    if (!this.img) return;

    global.addEventListener('resize', this.setParentNodeWidth);
    this.img.addEventListener('mousedown', this.setIsDynamicMapShowingTrue);
    this.img.addEventListener('mouseover', this.setIsMouseOverMapTrue);
    this.img.addEventListener('mouseout', this.setIsMouseOverMapFalse);
    this.setParentNodeWidth();
  };

  componentDidUpdate = (previousProps, previousState) =>
    !this.state.isDynamicMapShowing &&
    !previousState.isMouseOverMap &&
    this.state.isMouseOverMap &&
    global.setTimeout(
      () => this.state.isMouseOverMap && this.setIsDynamicMapShowingTrue(),
      500
    );

  componentWillUnmount = () => {
    if (!this.img) return;

    global.removeEventListener('resize', this.setParentNodeWidth);
    this.img.removeEventListener('mousedown', this.setIsDynamicMapShowingTrue);
    this.img.removeEventListener('mouseover', this.setIsMouseOverMapTrue);
    this.img.removeEventListener('mouseout', this.setIsMouseOverMapFalse);
  };

  createRef = img => (this.img = img);

  setParentNodeWidth = debounce(() => {
    const parentNodeWidth = getParentNodeWidth(this.img);

    this.setState({ parentNodeWidth });
  }, 10);

  setIsDynamicMapShowingTrue = () =>
    this.setState({ isDynamicMapShowing: true });

  setIsMouseOverMapTrue = () => this.setIsMouseOverMap(true);

  setIsMouseOverMapFalse = () => this.setIsMouseOverMap(false);

  setIsMouseOverMap = isMouseOverMap => this.setState({ isMouseOverMap });

  render = () => {
    const {
      apiKey,
      isFullBleed,
      isFluid,
      isShowingExactLocation,
      isShowingApproximateLocation,
      latitude,
      longitude,
      markers,
    } = this.props;

    const height = getHeight(this.props.height, isFullBleed, isFluid);

    return isFullBleed ? (
      <ReactGoogleMap
        apiKey={apiKey}
        height={height}
        isShowingApproximateLocation={isShowingApproximateLocation}
        isShowingExactLocation={isShowingExactLocation}
        latitude={latitude}
        longitude={longitude}
        markers={markers}
      />
    ) : (
      <Card fluid style={{ height, maxWidth: getMaxWidth(isFluid) }}>
        {isFluid || this.state.isDynamicMapShowing ? (
          <ReactGoogleMap
            apiKey={apiKey}
            height={height}
            isShowingApproximateLocation={isShowingApproximateLocation}
            isShowingExactLocation={isShowingExactLocation}
            latitude={latitude}
            longitude={longitude}
            markers={markers}
          />
        ) : (
          <img
            ref={this.createRef}
            src={getImgSrc(
              apiKey,
              height,
              isShowingApproximateLocation,
              isShowingExactLocation,
              latitude,
              longitude,
              this.state.parentNodeWidth
            )}
          />
        )}
      </Card>
    );
  };
}

Component.displayName = 'GoogleMap';

Component.defaultProps = {
  apiKey: GOOGLE_MAPS_API_KEY,
  height: '400px',
  isFluid: false,
  isFullBleed: false,
  isShowingExactLocation: false,
  isShowingApproximateLocation: false,
  latitude: 0,
  longitude: 0,
  markers: [],
};

Component.propTypes = {
  /** An [API key](https://developers.google.com/maps/documentation/javascript/get-api-key) for using Google Maps. */
  apiKey: PropTypes.string,
  /** A valid CSS value to set the height of the map. */
  height: PropTypes.string,
  /** Does the map show on a card which fills the width and height of its container. */
  isFluid: PropTypes.bool,
  /** Does the map fill the width and height of its container. */
  isFullBleed: PropTypes.bool,
  /** Is the map showing a marker for the approximate location at the center of the map. */
  isShowingApproximateLocation: PropTypes.bool,
  /** Is the map showing a marker for the exact location location at the center of the map. */
  isShowingExactLocation: PropTypes.bool,
  /** The latitude coordinate for the center of the map. */
  latitude: PropTypes.number,
  /** The longitude coordinate for the center of the map. */
  longitude: PropTypes.number,
  /** The markers to display on the map. */
  markers: PropTypes.arrayOf(
    PropTypes.shape({
      /** The React component to show as a marker. */
      component: PropTypes.node.isRequired,
      /**
       * A function called to determine the offset of the marker relative to its top left corner.
       * @param   {number} markerPixelWidth
       * @param   {number} markerPixelHeight
       * @return  {object}
       */
      getOffset: PropTypes.func,
      /** The latitude coordinate for the marker. */
      latitude: PropTypes.number,
      /** The longitude coordinate for the marker. */
      longitude: PropTypes.number,
    })
  ),
};
