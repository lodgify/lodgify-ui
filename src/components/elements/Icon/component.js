import React from 'react';
import PropTypes from 'prop-types';
import getClassNames from 'classnames';

import { Paragraph } from 'typography/Paragraph';

import { ICON_NAMES_LIST } from './constants';
import { getPath } from './utils/getPath';

/**
 * An icon is a glyph used to represent something else.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  color,
  hasBorder,
  isCircular,
  isColorInverted,
  isDisabled,
  isLabelLeft,
  labelText,
  labelWeight,
  name,
  path,
  size,
  ...otherProps
}) => (
  <i
    className={getClassNames('icon', color, size, {
      circular: isCircular,
      'has-border': hasBorder,
      'inverted grey': isDisabled,
      'has-label': !!labelText,
      inverted: isColorInverted,
    })}
    // Passing through props is required for
    // compatibility with dynamic components e.g. `ToolTip`
    {...otherProps}
  >
    {!!labelText && isLabelLeft && (
      <Paragraph weight={labelWeight}>{labelText}</Paragraph>
    )}
    <svg viewBox="0 0 24 24">
      <path d={getPath(name, path)} fill="currentColor" />
    </svg>
    {!!labelText && !isLabelLeft && (
      <Paragraph weight={labelWeight}>{labelText}</Paragraph>
    )}
  </i>
);

Component.displayName = 'Icon';

Component.defaultProps = {
  color: null,
  hasBorder: false,
  isCircular: false,
  isColorInverted: false,
  isDisabled: false,
  isLabelLeft: false,
  labelText: null,
  labelWeight: null,
  name: null,
  path: null,
  size: null,
};

Component.propTypes = {
  /** The color of the icon. */
  color: PropTypes.oneOf([
    'red',
    'orange',
    'yellow',
    'olive',
    'green',
    'teal',
    'blue',
    'violet',
    'purple',
    'pink',
    'brown',
    'grey',
    'light grey',
    'black',
  ]),
  /** Is the icon formatted to have a border  */
  hasBorder: PropTypes.bool,
  /** Is the icon formatted to appear circular  */
  isCircular: PropTypes.bool,
  /** Is the color of the icon inverted for contrast */
  isColorInverted: PropTypes.bool,
  /** Is the icon disabled. */
  isDisabled: PropTypes.bool,
  /** Is the label on the left hand side of the icon */
  isLabelLeft: PropTypes.bool,
  /** A visible label to display with the icon. */
  labelText: PropTypes.string,
  /** The weight of the label. */
  labelWeight: PropTypes.oneOf(['heavy', 'light']),
  /**
   * The name of the icon to display. Takes priority over `props.path`.
   * [See here for the full list of valid icon names](https://github.com/lodgify/lodgify-ui/blob/production/src/components/elements/Icon/constants.js)
   */
  name: PropTypes.oneOf(ICON_NAMES_LIST),
  /**
   * A string containing a series of path descriptions. Should fit a 24x24 viewBox.
   * [See MDN for documentation.](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d)
   */
  path: PropTypes.string,
  /** The size of the icon. */
  size: PropTypes.oneOf([
    'mini',
    'tiny',
    'small',
    'large',
    'big',
    'huge',
    'massive',
  ]),
};
