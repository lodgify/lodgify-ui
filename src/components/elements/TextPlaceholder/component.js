import React from 'react';
import { PropTypes } from 'prop-types';
import { Placeholder } from 'semantic-ui-react';

/**
 * A TextPlaceholder indicates that text content will soon appear in a layout.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ isFluid, length }) => (
  <Placeholder fluid={isFluid}>
    <Placeholder.Header>
      <Placeholder.Line length={length} />
    </Placeholder.Header>
  </Placeholder>
);

Component.displayName = 'TextPlaceholder';

Component.defaultProps = {
  isFluid: true,
  length: 'full',
};

Component.propTypes = {
  /** The placeholder fills the width of its parent. */
  isFluid: PropTypes.bool,
  /** The length of the lines in the placeholder. */
  length: PropTypes.oneOf([
    'very short',
    'short',
    'medium',
    'long',
    'very long',
    'full',
  ]),
};
