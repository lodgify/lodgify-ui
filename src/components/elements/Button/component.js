import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Button } from 'semantic-ui-react';

import { Icon } from 'elements/Icon';

/**
 * A button indicates a possible user action.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  children,
  isDisabled,
  isLoading,
  isPositionedRight,
  onClick,
  isRounded,
  isSecondary,
  isCompact,
  hasShadow,
  icon,
  size,
}) => (
  <Button
    circular={isRounded}
    className={cx({
      'has-shadow': !!hasShadow,
    })}
    compact={isCompact}
    disabled={isDisabled}
    floated={isPositionedRight ? 'right' : 'left'}
    loading={isLoading}
    onClick={onClick}
    secondary={isSecondary}
    size={size}
  >
    {!!icon && <Icon name={icon} />}
    {children}
  </Button>
);

Component.displayName = 'Button';

Component.defaultProps = {
  children: null,
  onClick: Function.prototype,
  isDisabled: false,
  isPositionedRight: false,
  isLoading: false,
  isRounded: false,
  isSecondary: false,
  isCompact: false,
  hasShadow: false,
  icon: null,
  size: null,
};

Component.propTypes = {
  /** The content to display in the button. */
  children: PropTypes.node,
  /** Has the button shadow. */
  hasShadow: PropTypes.bool,
  /** An icon to display in the button
   * [See here for the full list of valid icon names](https://github.com/lodgify/lodgify-ui/blob/master/src/components/elements/Icon/constants.js)
   */
  icon: PropTypes.string,
  /** Has the button got reduced padding. */
  isCompact: PropTypes.bool,
  /** Is the button disabled. */
  isDisabled: PropTypes.bool,
  /** Is the button in loading state. */
  isLoading: PropTypes.bool,
  /** Is the button positioned on the right hand side of its container. */
  isPositionedRight: PropTypes.bool,
  /** Has the button rounded borders. */
  isRounded: PropTypes.bool,
  /** Is the button secondary. */
  isSecondary: PropTypes.bool,
  /** The function to call when the button is clicked.
   *  @param {Object} event
   */
  onClick: PropTypes.func,
  /** The size of the button */
  size: PropTypes.string,
};
