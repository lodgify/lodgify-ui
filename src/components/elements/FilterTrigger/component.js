import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { Icon } from 'elements/Icon';
import { MORE_FILTERS } from 'utils/default-strings';

/**
 * A filter trigger removes filters and displays the available filters.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  onFiltersClick,
  onClearFiltersClick,
  filterTriggerText,
  hasFiltersSelected,
}) => (
  <div
    className={classNames('filter-trigger', {
      'has-filter': !!hasFiltersSelected,
    })}
  >
    <button onClick={onFiltersClick}>
      <Icon labelText={filterTriggerText} name="filter" />
    </button>
    {hasFiltersSelected && (
      <button onClick={onClearFiltersClick}>
        <Icon name="close" />
      </button>
    )}
  </div>
);

Component.displayName = 'FilterTrigger';

Component.defaultProps = {
  filterTriggerText: MORE_FILTERS,
  hasFiltersSelected: false,
  onClearFiltersClick: Function.prototype,
  onFiltersClick: Function.prototype,
};

Component.propTypes = {
  /** The text inside the filter trigger. */
  filterTriggerText: PropTypes.string,
  /** Filters are selected. */
  hasFiltersSelected: PropTypes.bool,
  /** The function to call when the clear filters button is clicked.
   *  @param {Object} event
   */
  onClearFiltersClick: PropTypes.func,
  /** The function to call when the filters button is clicked.
   *  @param {Object} event
   */
  onFiltersClick: PropTypes.func,
};
