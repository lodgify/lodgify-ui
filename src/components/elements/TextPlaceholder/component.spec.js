import React from 'react';
import { mount } from 'enzyme';
import { expectComponentToHaveDisplayName } from '@lodgify/enzyme-jest-expect-helpers';

import { Component as TextPlaceholder } from './component';

const getTextPlaceholder = () => mount(<TextPlaceholder />);

describe('<TextPlaceholder />', () => {
  it('should render a single Semantic UI `TextPlaceholder` component', () => {
    const actual = getTextPlaceholder();

    expect(actual).toMatchSnapshot();
  });

  it('should have displayName `TextPlaceholder`', () => {
    expectComponentToHaveDisplayName(TextPlaceholder, 'TextPlaceholder');
  });
});
