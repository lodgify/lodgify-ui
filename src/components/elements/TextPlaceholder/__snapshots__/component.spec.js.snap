// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextPlaceholder /> should render a single Semantic UI \`TextPlaceholder\` component 1`] = `
<TextPlaceholder
  isFluid={true}
  length="full"
>
  <Placeholder
    fluid={true}
  >
    <div
      className="ui fluid placeholder"
    >
      <PlaceholderHeader>
        <div
          className="header"
        >
          <PlaceholderLine
            length="full"
          >
            <div
              className="line full"
            />
          </PlaceholderLine>
        </div>
      </PlaceholderHeader>
    </div>
  </Placeholder>
</TextPlaceholder>
`;
